dynamiqs/integrators/mcsolve/diffrax_integrator.py

```diff
@@ -0,0 +1,258 @@
+                # in diffrax/_integrate.py which can be found here https://github.com/patrick-kidger/diffrax/blob/main/diffrax/_integrate.py#L427-L458  # noqa E501
+                def _inner_cond_fun(__save_state):
+                    return (self.ts[__save_state.save_index] <= final_time) & (
+                        __save_state.save_index < len(self.ts)
+                    )
+
+                def _inner_body_fun(__save_state):
+                    _t = self.ts[__save_state.save_index]
+                    _y = psi_interpolator.evaluate(_t)
+                    _ts = __save_state.ts.at[__save_state.save_index].set(_t)
+                    _ys = jtu.tree_map(
+                        lambda __y, __ys: __ys.at[__save_state.save_index].set(__y),
+                        self.save(unit(_y)),
+                        __save_state.ys,
+                    )
+                    return SaveState(
+                        ts=_ts,
+                        ys=_ys,
+                        save_index=__save_state.save_index + 1,
+                        jump_times=__save_state.jump_times,
+                    )
+
+                save_state = while_loop(
+                    _inner_cond_fun,
+                    _inner_body_fun,
+                    _save_state,
+                    max_steps=len(self.ts),
+                    buffers=_inner_buffers,
+                    kind='checkpointed',
+                )
+                return save_state
+
+            save_state = save_ts_and_states(_state.save_state)
+            jump_times = save_state.jump_times.at[_state.num_jumps].set(jump_time)
+            save_state = SaveState(
+                ts=save_state.ts,
+                ys=save_state.ys,
+                save_index=save_state.save_index,
+                jump_times=jump_times,
+            )
+            return State(
+                save_state=save_state,
+                final_state=final_state,
+                final_time=final_time,
+                num_jumps=_state.num_jumps + 1,
+                key=_next_key,
+            )
+
+        return while_loop(
+            _outer_cond_fun,
+            _outer_body_fun,
+            state,
+            max_steps=self.options.max_jumps,
+            buffers=_outer_buffers,
+            kind='checkpointed',
+        )
+
+    def _sample_jump_ops(self, t: Array, psi: Array, key: Array) -> Array:
+        """given a state psi at time t that should experience a jump,
+        randomly sample one jump operator from among the provided jump_ops.
+        The probability that a certain jump operator is selected is weighted
+        by the probability that such a jump can occur. For instance for a qubit
+        experiencing amplitude damping, if it is in the ground state then
+        there is probability zero of experiencing an amplitude damping event.
+        """
+        Ls = stack([L(t) for L in self.Ls])
+        Lsd = dag(Ls)
+        probs = expect(Lsd @ Ls, psi)
```

Comment:
Re your comment in vector field, should we do soemthing like
```suggestion
        Ls = stack([L(t) for L in self.Ls])
        Ls_on_psi = [L @ psi for L in Ls]
        probs = jnp.asarray([_psi.dag() @ _psi for _psi in Ls_on_psi]).squeeze(-1)
```
so that we only have matrix-vector products instead of the current matrix-matrix product?

I think though you are pointing out something different, that is the use of stack: if I just use a listcomp, I get an error when I try to sample the indicated jump operator
```    
return Ls[sample_idx]
jax.errors.TracerIntegerConversionError: The __index__() method was called on traced array with shape int32[].
The error occurred while tracing the function _fn at /Users/danielweiss/PycharmProjects/equinox/equinox/_make_jaxpr.py:36 for jit. This concrete value was not available in Python because it depends on the values of the arguments _dynamic_flat[8] and _dynamic_flat[11].
```
So I'm not sure how to proceed without using `stack`